fabric-biomes-v1/src/testmod/java/net/fabricmc/fabric/test/biome/FabricBiomeTest.java

```diff
@@ -16,26 +16,52 @@
+import net.minecraft.util.registry.RegistryKey;
 import net.minecraft.world.biome.Biome;
-import net.minecraft.world.biome.Biomes;
+import net.minecraft.world.biome.BuiltInBiomes;
+import net.minecraft.world.biome.DefaultBiomeCreator;
 
 import net.fabricmc.api.ModInitializer;
 import net.fabricmc.fabric.api.biomes.v1.NetherBiomes;
 import net.fabricmc.fabric.api.biomes.v1.OverworldBiomes;
+import net.fabricmc.fabric.api.biomes.v1.OverworldClimate;
 
+/**
+ * <b>NOTES FOR TESTING:</b>
+ * When running with this test-mod, also test this when running a dedicated server since there
+ * are significant differences between server + client and how they sync biomes.
+ *
+ * <p>Ingame, you can use <code>/locatebiome</code> since we use nether- and end-biomes in the overworld,
+ * and vice-versa, making them easy to find to verify the injection worked.
+ *
+ * <p>If you don't find a biome right away, teleport far away (~10000 blocks) from spawn and try again.
+ */
 public class FabricBiomeTest implements ModInitializer {
 	public static final String MOD_ID = "fabric-biome-api-v1-testmod";
 
-	@Override public void onInitialize() {
-		Biome biome = Registry.register(BuiltinRegistries.BIOME, new Identifier(MOD_ID, "test_crimson_forest"), DefaultBiomeCreator.createCrimsonForest());
-		NetherBiomes.addNetherBiome(Biomes.BEACH, new Biome.MixedNoisePoint(0.0F, 0.5F, 0.0F, 0.0F, 0.1F));
-		NetherBiomes.addNetherBiome(biome, new Biome.MixedNoisePoint(0.0F, 0.5F, 0.0F, 0.0F, 0.275F));
+	private static final RegistryKey<Biome> TEST_CRIMSON_FOREST = RegistryKey.of(Registry.BIOME_KEY, new Identifier(MOD_ID, "test_crimson_forest"));
+
+	private static final RegistryKey<Biome> CUSTOM_PLAINS = RegistryKey.of(Registry.BIOME_KEY, new Identifier(MOD_ID, "custom_plains"));
+
+	@Override
+	public void onInitialize() {
+		Registry.register(BuiltinRegistries.BIOME, TEST_CRIMSON_FOREST.getValue(), DefaultBiomeCreator.createCrimsonForest());
+
+		NetherBiomes.addNetherBiome(BuiltInBiomes.BEACH, new Biome.MixedNoisePoint(0.0F, 0.5F, 0.0F, 0.0F, 0.1F));
+		NetherBiomes.addNetherBiome(TEST_CRIMSON_FOREST, new Biome.MixedNoisePoint(0.0F, 0.5F, 0.0F, 0.0F, 0.275F));
+
+		Registry.register(BuiltinRegistries.BIOME, CUSTOM_PLAINS.getValue(), DefaultBiomeCreator.createPlains(false));
+		OverworldBiomes.addBiomeVariant(BuiltInBiomes.PLAINS, CUSTOM_PLAINS, 1);
+
+		OverworldBiomes.addEdgeBiome(BuiltInBiomes.PLAINS, BuiltInBiomes.END_BARRENS, 0.9);
+
+		OverworldBiomes.addShoreBiome(BuiltInBiomes.FOREST, BuiltInBiomes.NETHER_WASTES, 0.9);
+
+		OverworldBiomes.addHillsBiome(BuiltInBiomes.BAMBOO_JUNGLE, BuiltInBiomes.BASALT_DELTAS, 0.9);
 
-		Biome customPlains = Registry.register(BuiltinRegistries.BIOME, new Identifier(MOD_ID, "custom_plains"), DefaultBiomeCreator.createPlains(null, false));
-		OverworldBiomes.addBiomeVariant(Biomes.PLAINS, customPlains, 1);
+		OverworldBiomes.addContinentalBiome(BuiltInBiomes.END_HIGHLANDS, OverworldClimate.DRY, 0.5);
 	}
```

Comment:
Would be nice if we could get this to work using json biomes, even if you think we shouldn't use those.
Putting this here cause it'd be part of the test mod.

Reply:
I don't think defining the same thing you could define with a constructor call in a JSON file justifies the added complexity for this API and everyone else involved. There is literally nothing else you gain from doing it in a JSON file.

Datapacks don't factor into this. You can _still_ override biome definitions from data packs like you can with Vanilla biomes, and defining a _new_ biome in a datapack ist pointless since the mod that calls this API could not know about it beforehand.